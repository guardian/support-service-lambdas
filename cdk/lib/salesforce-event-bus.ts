import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import { type App, Duration } from 'aws-cdk-lib';

import { nodeVersion } from './node-version';

export class SalesforceEventBus extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const app = 'salesforce-event-bus';

		new GuLambdaFunction(this, 'SalesforceEventBusPlaceholder', {
			app,
			memorySize: 1024,
			fileName: `${app}.zip`,
			runtime: nodeVersion,
			timeout: Duration.seconds(300),
			handler: 'salesforceEventBusPlaceholder.handler',
			functionName: `salesforce-event-bus-placeholder-${this.stage}`,
		});
	}
}
