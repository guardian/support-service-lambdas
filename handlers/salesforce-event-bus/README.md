## Salesforce Event Bus

### Description

An Eventbridge Event Bus used to relay messages from Salesforce to AWS Services. This uses [Salesforce Event Relay](https://developer.salesforce.com/docs/platform/event-relay/overview) which streams events to AWS natively without the need to build custom middleware. Event Relay leverages the Eventbridge Partner APIs to connect securely without the need to share any credentials.

This should be used for event-driven, asynchronous integrations between Salesforce and AWS.

#### Dead Letter Queue

A generic dead letter queue `dead-letters-salesforce-event-bus-queue-<STAGE>` is implemented that can be used by all rules. This is only intended to be used by the rules should they fail to relay events to the intended target(s), it is separate from dead letter queues used by any targets.

When Eventbridge sends events to this DLQ the attributes will include the error code and message, which can be used to determine the cause of the failure.

<img width="739" alt="Screenshot 2025-06-23 at 17 28 27" src="https://github.com/user-attachments/assets/0dac6e6a-8fc1-4ccc-93da-8cd505cc0103" />

An alarm is activated when there is at least one item in the DLQ.

### Implementation Steps

####Â Salesforce

All pre-requisite steps are described [here](https://help.salesforce.com/s/articleView?id=platform.ev_relay_create_prereqs.htm&type=5) - the Named Credential, Event Relay and Event Channel (`AWS_Platform_Event_Bus__chn`) have already been created and activated.

To add a new type of Event:

1. [Create a new Platform Event](https://help.salesforce.com/s/articleView?id=platform.ev_relay_define_pe.htm&type=5) (if required)
   a. Implement automation to publish a Platform Event when required
2. [Add the Platform Event](https://help.salesforce.com/s/articleView?id=platform.ev_relay_create_channel_member_pe.htm&type=5) to the existing `AWS_Platform_Event_Bus__chn` Channel

#### AWS Eventbridge

Create a new Eventbridge rule to forward the event to the relevant AWS service.

The rule _must_ include filtering to ensure that only relevant events are processed. This should at least specify the `detail-type` which corresponds to the name of the Platform Event in Salesforce, but could include additional information.

### Implemented Rules

#### Contact Update to SQS

This rule entirely replaces [salesforce-message-handler](https://github.com/guardian/salesforce-message-handler/tree/main) that previously used Salesforce's legacy workflow automation with outbound messages to sync data to other systems after Contact record updates (including Zuora and Identity).

The destination queue is `salesforce-outbound-messages-<STAGE>` which is consumed by membership-workflow. It receives a Salesforce Record Id from an updated Contact, queries the Contact in Salesforce for the latest data and updates other systems.
