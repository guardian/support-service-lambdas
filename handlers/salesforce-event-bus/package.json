{
	"name": "salesforce-event-bus",
	"description": "An Eventbridge event bus to receive and send message to and from Salesforce",
	"scripts": {
		"test": "jest --group=-integration",
		"it-test": "jest --group=integration",
		"build": "esbuild --bundle --platform=node --target=node20 --outdir=target src/handlers/*.ts",
		"lint": "eslint src/**/*.ts test/**/*.ts",
		"package": "pnpm type-check && pnpm check-formatting && pnpm test && pnpm build && cd target && zip -qr salesforce-event-bus.zip ./*.js",
		"type-check": "tsc --noEmit",
		"check-formatting": "prettier --check **.ts",
		"fix-formatting": "prettier --write **.ts"
	},
	"dependencies": {}
}
