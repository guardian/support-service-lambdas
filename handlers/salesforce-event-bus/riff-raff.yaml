stacks:
  - support
regions:
  - eu-west-1
allowedStages:
  - CODE
  - PROD
deployments:
  salesforce-event-bus-cloudformation:
    type: cloud-formation
    app: salesforce-event-bus
    parameters:
      templateStagePaths:
        CODE: salesforce-event-bus-CODE.template.json
        PROD: salesforce-event-bus-PROD.template.json
