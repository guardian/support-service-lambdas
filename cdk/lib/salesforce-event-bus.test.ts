import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SalesforceEventBus } from './salesforce-event-bus';

describe('The SalesforceEventBus stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const codeStack = new SalesforceEventBus(app, `salesforce-event-bus-CODE`, {
			stack: 'membership',
			stage: 'CODE',
		});
		const prodStack = new SalesforceEventBus(app, `salesforce-event-bus-PROD`, {
			stack: 'membership',
			stage: 'PROD',
		});
		expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
		expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
	});
});
