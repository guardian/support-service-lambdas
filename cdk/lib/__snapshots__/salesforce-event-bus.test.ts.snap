// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SalesforceEventBus stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "SalesforceEventBusPlaceholder69393C1F": {
      "DependsOn": [
        "SalesforceEventBusPlaceholderServiceRoleDefaultPolicy699BD152",
        "SalesforceEventBusPlaceholderServiceRoleA0FFC02B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "membership/CODE/salesforce-event-bus/salesforce-event-bus.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "salesforce-event-bus",
            "STACK": "membership",
            "STAGE": "CODE",
          },
        },
        "FunctionName": "salesforce-event-bus-placeholder-CODE",
        "Handler": "salesforceEventBusPlaceholder.handler",
        "LoggingConfig": {
          "LogFormat": "JSON",
        },
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "SalesforceEventBusPlaceholderServiceRoleA0FFC02B",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "salesforce-event-bus",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SalesforceEventBusPlaceholderServiceRoleA0FFC02B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "salesforce-event-bus",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SalesforceEventBusPlaceholderServiceRoleDefaultPolicy699BD152": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/membership/CODE/salesforce-event-bus/salesforce-event-bus.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/membership/salesforce-event-bus",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/membership/salesforce-event-bus/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SalesforceEventBusPlaceholderServiceRoleDefaultPolicy699BD152",
        "Roles": [
          {
            "Ref": "SalesforceEventBusPlaceholderServiceRoleA0FFC02B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SfBusToContactUpdateQueue75131C06": {
      "Properties": {
        "Description": "Send an SF Contact Update event from the SF bus to the salesforce-outbound-messages-[STAGE] SQS queue for consumption by membership-workflow",
        "EventBusName": "aws.partner/salesforce.com/00D9E0000004jvhUAA/0YLUD00000008Ll4AI",
        "EventPattern": {
          "detail-type": [
            "Contact_Update__e",
          ],
          "source": [
            {
              "prefix": "aws.partner/salesforce.com",
            },
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": "arn:aws:sqs:eu-west-1:865473395570:salesforce-outbound-messages-CODE",
            "DeadLetterConfig": {
              "Arn": {
                "Fn::GetAtt": [
                  "deadletterssalesforceeventbusqueue29BFA779",
                  "Arn",
                ],
              },
            },
            "Id": "Target0",
            "RetryPolicy": {
              "MaximumEventAgeInSeconds": 7200,
              "MaximumRetryAttempts": 2,
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "deadletterssalesforceeventbusqueue29BFA779": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "dead-letters-salesforce-event-bus-queue-CODE",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "deadletterssalesforceeventbusqueuePolicy0CE2F482": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "SfBusToContactUpdateQueue75131C06",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "deadletterssalesforceeventbusqueue29BFA779",
                  "Arn",
                ],
              },
              "Sid": "AllowEventRulesalesforceeventbusCODESfBusToContactUpdateQueueFE76F3CE",
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "deadletterssalesforceeventbusqueue29BFA779",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
  },
}
`;

exports[`The SalesforceEventBus stack matches the snapshot 2`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "SalesforceEventBusPlaceholder69393C1F": {
      "DependsOn": [
        "SalesforceEventBusPlaceholderServiceRoleDefaultPolicy699BD152",
        "SalesforceEventBusPlaceholderServiceRoleA0FFC02B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "membership/PROD/salesforce-event-bus/salesforce-event-bus.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "salesforce-event-bus",
            "STACK": "membership",
            "STAGE": "PROD",
          },
        },
        "FunctionName": "salesforce-event-bus-placeholder-PROD",
        "Handler": "salesforceEventBusPlaceholder.handler",
        "LoggingConfig": {
          "LogFormat": "JSON",
        },
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "SalesforceEventBusPlaceholderServiceRoleA0FFC02B",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "salesforce-event-bus",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SalesforceEventBusPlaceholderServiceRoleA0FFC02B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "salesforce-event-bus",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SalesforceEventBusPlaceholderServiceRoleDefaultPolicy699BD152": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/membership/PROD/salesforce-event-bus/salesforce-event-bus.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/membership/salesforce-event-bus",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/membership/salesforce-event-bus/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SalesforceEventBusPlaceholderServiceRoleDefaultPolicy699BD152",
        "Roles": [
          {
            "Ref": "SalesforceEventBusPlaceholderServiceRoleA0FFC02B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SfBusToContactUpdateQueue75131C06": {
      "Properties": {
        "Description": "Send an SF Contact Update event from the SF bus to the salesforce-outbound-messages-[STAGE] SQS queue for consumption by membership-workflow",
        "EventBusName": "aws.partner/salesforce.com/00D20000000nq5gEAA/0YLQv00000000zJOAQ",
        "EventPattern": {
          "detail-type": [
            "Contact_Update__e",
          ],
          "source": [
            {
              "prefix": "aws.partner/salesforce.com",
            },
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": "arn:aws:sqs:eu-west-1:865473395570:salesforce-outbound-messages-PROD",
            "DeadLetterConfig": {
              "Arn": {
                "Fn::GetAtt": [
                  "deadletterssalesforceeventbusqueue29BFA779",
                  "Arn",
                ],
              },
            },
            "Id": "Target0",
            "RetryPolicy": {
              "MaximumEventAgeInSeconds": 7200,
              "MaximumRetryAttempts": 2,
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "deadletterssalesforceeventbusqueue29BFA779": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "dead-letters-salesforce-event-bus-queue-PROD",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "deadletterssalesforceeventbusqueuePolicy0CE2F482": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Fn::GetAtt": [
                      "SfBusToContactUpdateQueue75131C06",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "deadletterssalesforceeventbusqueue29BFA779",
                  "Arn",
                ],
              },
              "Sid": "AllowEventRulesalesforceeventbusPRODSfBusToContactUpdateQueueC08A5AB1",
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "deadletterssalesforceeventbusqueue29BFA779",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
  },
}
`;
